crates/pulsar-core/src/pdk/module.rs

```diff
@@ -31,94 +33,140 @@ pub trait PulsarModule: Send {
+/// A simpler version of [`PulsarModule`] which is often enough. A blanket implementation ensures that
+/// [`PulsarModule`] is implemented for all implementors of [`BasicPulsarModule`].
+pub trait BasicPulsarModule: Send + Sync {
+    type Config: for<'a> TryFrom<&'a ModuleConfig, Error = ConfigError> + Send + Sync + 'static;
+    type State: Send + 'static;
 
-    fn init_extra_state(
-        module: &T,
-        config: &T::Config,
-    ) -> impl Future<Output = Result<Self::ExtraState, ModuleError>> + Send;
+    const MODULE_NAME: &'static str;
+    const DEFAULT_ENABLED: bool;
 
-    fn trigger(
-        extra_state: &mut Self::ExtraState,
-    ) -> impl Future<Output = Result<Self::TriggerOutput, ModuleError>> + Send;
+    fn init_state(
+        &self,
+        config: &Self::Config,
+        ctx: &ModuleContext,
+    ) -> impl Future<Output = Result<Self::State, ModuleError>> + Send;
 
-    fn action(
-        trigger_output: &Self::TriggerOutput,
-        config: &T::Config,
-        state: &mut T::State,
+    fn on_config_change(
+        _new_config: &Self::Config,
+        _state: &mut Self::State,
         ctx: &ModuleContext,
-    ) -> impl Future<Output = Result<(), ModuleError>> + Send;
+    ) -> impl Future<Output = Result<(), ModuleError>> + Send {
+        ctx.stop_cfg_recv()
+    }
+
+    fn on_event(
+        _event: &Event,
+        _config: &Self::Config,
+        _state: &mut Self::State,
+        ctx: &ModuleContext,
+    ) -> impl Future<Output = Result<(), ModuleError>> + Send {
+        ctx.stop_event_recv()
+    }
+
+    /// Default: normal state drop
+    fn graceful_stop(_state: Self::State) -> impl Future<Output = Result<(), ModuleError>> + Send {
+        std::future::ready(Ok(()))
+    }
 }
 
-pub struct NoExtra(());
+impl<T> PulsarModule for T
+where
+    T: BasicPulsarModule,
+{
+    type Config = T::Config;
+
+    type State = T::State;
 
-impl<T: PulsarModule> Extra<T> for NoExtra {
-    type ExtraState = ();
+    type Extension = ();
 
     type TriggerOutput = ();
 
-    fn init_extra_state(
-        module: &T,
-        config: &<T as PulsarModule>::Config,
-    ) -> impl Future<Output = Result<Self::ExtraState, ModuleError>> + Send {
-        let _ = module;
-        let _ = config;
-        std::future::ready(Ok(()))
+    const MODULE_NAME: &'static str = Self::MODULE_NAME;
+
+    const DEFAULT_ENABLED: bool = Self::DEFAULT_ENABLED;
+
+    async fn init_state(
+        &self,
+        config: &Self::Config,
+        ctx: &ModuleContext,
+    ) -> Result<(Self::State, Self::Extension), ModuleError> {
+        BasicPulsarModule::init_state(self, config, ctx)
+            .await
+            .map(|v| (v, ()))
     }
 
-    fn trigger(
-        extra_state: &mut Self::ExtraState,
-    ) -> impl Future<Output = Result<Self::TriggerOutput, ModuleError>> {
-        let _ = extra_state;
-        std::future::pending()
+    async fn trigger(
+        _extra_state: &mut Self::Extension,
```

Comment:
There is still `extra_state` here. Can we rename to `state`?

Reply:
Yup, my bad. Fixed.